schema-engine/core/src/state.rs

```diff
@@ -27,7 +27,47 @@ pub(crate) struct EngineState {
     // - a full schema
     //
     // To a channel leading to a spawned MigrationConnector.
-    connectors: Mutex<HashMap<String, mpsc::Sender<ErasedConnectorRequest>>>,
+    connectors: Mutex<HashMap<ConnectorRequestType, mpsc::Sender<ErasedConnectorRequest>>>,
+}
+
+impl EngineState {
+    fn get_url_from_schemas(&self, container: &SchemasContainer) -> CoreResult<String> {
+        for path in &container.files {
+            if let Ok(url) = self.get_url_from_path(path) {
+                return Ok(url);
+            }
+        }
+
+        Err(ConnectorError::from_msg("No datasource block found".to_owned()))
+    }
+
+    fn get_url_from_path(&self, container: &SchemaContainer) -> CoreResult<String> {
+        let file_path = &container.path;
+        let mut schema_file = std::fs::File::open(file_path)
+            .map_err(|err| ConnectorError::from_source(err, "Opening Prisma schema file."))?;
+        let mut schema_string = String::new();
+        schema_file
+            .read_to_string(&mut schema_string)
+            .map_err(|err| ConnectorError::from_source(err, "Reading Prisma schema file."))?;
+        let (datasource, url, _, _) = crate::parse_configuration(&schema_string)?;
+
+        Ok(std::path::Path::new(file_path)
+            .parent()
+            .map(|config_dir| psl::set_config_dir(datasource.active_connector.flavour(), config_dir, &url).into_owned())
+            .unwrap_or(url))
+    }
```

Comment:
Refactor by using parse_configuration_multi. There should be no need to re-read the files.